ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/akkastreams/dispatcher/SignalDispatcher.scala

```diff
@@ -63,28 +62,35 @@ class SignalDispatcher private () {
   /** Closes this SignalDispatcher by gracefully completing the existing Source subscriptions.
     * For any downstream with pending signals, at least one such signal will be sent first.
     */
-  def shutdown(): Future[Unit] = shutdownInternal(_.complete())
+  def shutdown(): Future[Unit] =
+    shutdownInternal { source =>
+      source.complete()
+      source.watchCompletion()
+    }
 
   /** Closes this SignalDispatcher by failing the existing Source subscriptions with the provided throwable. */
-  def fail(throwable: Throwable): Future[Unit] =
-    shutdownInternal(_.fail(throwable)).transform {
-      // This throwable is expected so map to Success
-      case Failure(`throwable`) => Success(())
-      case other => other
-    }(ExecutionContext.parasitic)
+  def fail(newThrowable: () => Throwable): Future[Unit] =
+    shutdownInternal { source =>
+      implicit val ec: ExecutionContext = scala.concurrent.ExecutionContext.parasitic
+      val throwable = newThrowable()
+      source.fail(throwable)
+      source
+        .watchCompletion()
+        .recoverWith {
+          // This throwable is expected so map to Success
+          case `throwable` => Future.unit
```

Comment:
in order to not let the failure to bubble up if it's expected.

Reply:
it seems to me this code-path is executed as we reset the in-memory state.
In DispatcherState stopDispatcher we first switch to not running, log the error what is bubbling up, then fail the whole initialization, which triggers an indexer restart in 10 seconds, which initializes quickly (since we already forgot about the dispatcher which failed).

Since we are not doing much about such an error, just logging it (and it anyway can be several instead of one), and by emitting an error we are causing a non-necessary indexer restart (which is not really helping on anything this regard): can't we just log in here, and not letting errors bubble up at all?
Or if we are being serious about some countermeasures if API stream teardown failed, then we still need to implement them ;)

Reply:
Makes sense. Switching to log and continue (with success)